### ISSUE TITLE
Planning failure for non-equi left join

### ISSUE THREAD

**Issue by martint** [prestodb/presto#8673]
```sql
SELECT 1
FROM orders a
LEFT JOIN lineitem b ON a.comment >= b.comment
WHERE b.comment IS NOT NULL
```

Fails with:

```
java.lang.IllegalArgumentException: Invalid node. Predicate dependencies ([comment_1, comment]) not in source plan output ([comment_1])
	at com.google.common.base.Preconditions.checkArgument(Preconditions.java:146)
	at com.facebook.presto.sql.planner.sanity.ValidateDependenciesChecker.checkDependencies(ValidateDependenciesChecker.java:597)
	at com.facebook.presto.sql.planner.sanity.ValidateDependenciesChecker.access$100(ValidateDependenciesChecker.java:79)
	at com.facebook.presto.sql.planner.sanity.ValidateDependenciesChecker$Visitor.visitFilter(ValidateDependenciesChecker.java:214)
	at com.facebook.presto.sql.planner.sanity.ValidateDependenciesChecker$Visitor.visitFilter(ValidateDependenciesChecker.java:93)
	at com.facebook.presto.sql.planner.plan.FilterNode.accept(FilterNode.java:72)
	at com.facebook.presto.sql.planner.sanity.ValidateDependenciesChecker$Visitor.visitProject(ValidateDependenciesChecker.java:232)
	at com.facebook.presto.sql.planner.sanity.ValidateDependenciesChecker$Visitor.visitProject(ValidateDependenciesChecker.java:93)
	at com.facebook.presto.sql.planner.plan.ProjectNode.accept(ProjectNode.java:92)
	at com.facebook.presto.sql.planner.sanity.ValidateDependenciesChecker$Visitor.visitExchange(ValidateDependenciesChecker.java:449)
	at com.facebook.presto.sql.planner.sanity.ValidateDependenciesChecker$Visitor.visitExchange(ValidateDependenciesChecker.java:93)
	at com.facebook.presto.sql.planner.plan.ExchangeNode.accept(ExchangeNode.java:196)
	at com.facebook.presto.sql.planner.sanity.ValidateDependenciesChecker$Visitor.visitExchange(ValidateDependenciesChecker.java:449)
	at com.facebook.presto.sql.planner.sanity.ValidateDependenciesChecker$Visitor.visitExchange(ValidateDependenciesChecker.java:93)
	at com.facebook.presto.sql.planner.plan.ExchangeNode.accept(ExchangeNode.java:196)
	at com.facebook.presto.sql.planner.sanity.ValidateDependenciesChecker$Visitor.visitJoin(ValidateDependenciesChecker.java:307)
	at com.facebook.presto.sql.planner.sanity.ValidateDependenciesChecker$Visitor.visitJoin(ValidateDependenciesChecker.java:93)
	at com.facebook.presto.sql.planner.plan.JoinNode.accept(JoinNode.java:218)
	at com.facebook.presto.sql.planner.sanity.ValidateDependenciesChecker$Visitor.visitProject(ValidateDependenciesChecker.java:232)
	at com.facebook.presto.sql.planner.sanity.ValidateDependenciesChecker$Visitor.visitProject(ValidateDependenciesChecker.java:93)
	at com.facebook.presto.sql.planner.plan.ProjectNode.accept(ProjectNode.java:92)
	at com.facebook.presto.sql.planner.sanity.ValidateDependenciesChecker$Visitor.visitExchange(ValidateDependenciesChecker.java:449)
	at com.facebook.presto.sql.planner.sanity.ValidateDependenciesChecker$Visitor.visitExchange(ValidateDependenciesChecker.java:93)
	at com.facebook.presto.sql.planner.plan.ExchangeNode.accept(ExchangeNode.java:196)
	at com.facebook.presto.sql.planner.sanity.ValidateDependenciesChecker$Visitor.visitOutput(ValidateDependenciesChecker.java:276)
	at com.facebook.presto.sql.planner.sanity.ValidateDependenciesChecker$Visitor.visitOutput(ValidateDependenciesChecker.java:93)
	at com.facebook.presto.sql.planner.plan.OutputNode.accept(OutputNode.java:82)
	at com.facebook.presto.sql.planner.sanity.ValidateDependenciesChecker.validate(ValidateDependenciesChecker.java:90)
	at com.facebook.presto.sql.planner.sanity.ValidateDependenciesChecker.validate(ValidateDependenciesChecker.java:85)
	at com.facebook.presto.sql.planner.sanity.PlanSanityChecker.lambda$validateFinalPlan$0(PlanSanityChecker.java:54)
	at com.google.common.collect.ImmutableList.forEach(ImmutableList.java:408)
	at com.facebook.presto.sql.planner.sanity.PlanSanityChecker.validateFinalPlan(PlanSanityChecker.java:54)
	at com.facebook.presto.sql.planner.LogicalPlanner.plan(LogicalPlanner.java:140)
	at com.facebook.presto.sql.planner.LogicalPlanner.plan(LogicalPlanner.java:122)
	at com.facebook.presto.execution.SqlQueryExecution.doAnalyzeQuery(SqlQueryExecution.java:358)
	at com.facebook.presto.execution.SqlQueryExecution.analyzeQuery(SqlQueryExecution.java:337)
	at com.facebook.presto.execution.SqlQueryExecution.start(SqlQueryExecution.java:269)
	at com.facebook.presto.execution.QueuedExecution.lambda$start$1(QueuedExecution.java:62)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
```

The issue doesn't occur if the WHERE clause is removed, if the join is converted to an INNER JOIN or an equality join.

This seems to be a recent regression, although I haven't tracked down which commit introduced it.

**Comment by martint** [comment]
It seems to be caused by a6b2e75d3720a71dbb6e89eb2ef4f1633cce6e05.

cc @losipiuk

**Comment by losipiuk** [comment]
The problem appears where we actually extract the `FilterNode` out of the `JoinNode`'s `filterFunction`.
We get the failure if new `FilterNode` requires symbols which were not needed by nodes above `JoinNode` in original plan tree (an therefore not part of `JoinNode.getOutputSymbols()`.

**Comment by losipiuk** [comment]
@martint, @kokosing  please take a look at proposed solution in https://github.com/prestodb/presto/pull/8682.

It simplifies the logic in `PredicatePushDown` concerning maintaining outputSymbols in produced by `JoinNode`. 
We just produce everything from `JoinNode` and trim it with `ProjectNode` always.

The other "cleanup" optimizers (`UnaliasSymbolRefereces`, `PruneUnreferencedOutputs`, `InlineProjections`, `RemoveRedundantIdentityProjections`) should remove added unnecessary projections afterwards.

It generally happens, but possibly there are some misbehaviours. One I observed is for following query:

```
explain SELECT 1 FROM orders a LEFT JOIN lineitem b ON a.comment >= b.comment WHERE b.comment IS NOT NULL;
```

In plan we get two projections one below the other:
```
      - RemoteExchange[GATHER] => expr:integer
             Cost: {rows: 30000000, bytes: ?}
         - Project[] => [expr:integer]
                 Cost: {rows: 30000000, bytes: ?}
                 expr := 1
             - FilterProject[filterPredicate = ("comment" >= CAST("comment_1" AS varchar(79)))] => []
                     Cost: {rows: 30000000, bytes: ?}/{rows: 30000000, bytes: ?}
```
The lower one being empty. It probably is not a performance killer but sth which should be addressed.

If I replace `1` with above query with any column it looks good:

```
explain SELECT a.orderdate FROM orders a LEFT JOIN lineitem b ON a.comment >= b.comment WHERE b.comment IS NOT NULL;
```
is planned as
```
     - RemoteExchange[GATHER] => orderdate:date
             Cost: {rows: 30000000, bytes: ?}
         - FilterProject[filterPredicate = ("comment" >= CAST("comment_1" AS varchar(79)))] => [orderdate:date]
                 Cost: {rows: 30000000, bytes: ?}/{rows: 30000000, bytes: ?}
```

What do you think about such approach. I feel like we should keep individual optimizers as simple as possible and not repeat the work which is done in other places.

**Comment by findepi** [comment]
Fixed by #8682
